### PR TITLE
Porting the codebase to Python 3

### PR DESCRIPTION
**Pull request by jonas-schmitt**
Since Python 2.7 will not be maintained past 2020, it makes sense to switch to Python 3 as early as possible. This also fixes potential compatibility problems with Python 3 libraries. I ported the complete Python-codebase to Python 3, which could be mostly automated using 2to3 and by providing the -py3 flag to swig. The only exception was the PyNdRangeIterator (defined in NdRange.i), where the next() method had to be adapted to make the generated Python class compatible with Python 3.

All test seem to produce the correct output, except the one broken, mentioned in the current issue. I still invite you to carefully look at the changes and apply some additional testing.

There currently seems to be a weird problem with the Python 3 installation used in Travis CI. The build  works fine on my machine (Ubuntu 16.04.4 LTS, Python 3.6.4).

### CODE REVIEW

File: demo/block_jac_2x2.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 I = operator.identity(grid)
 E = (I - omega * D.inverse() * A)
 
-print(E.symbol().spectral_radius())
+print((E.symbol().spectral_radius()))
```
Comment (hrittich) [inline demo/block_jac_2x2.py]: These additional parenthesis are unecessary.

File: demo/block_rb_jac_2x2.py [hunk 2]
```diff
@@ -39,7 +39,7 @@
 # Define the RB-Jacobi error propagator
 E = (FR + FB * J) * (FB + FR * J)
 
-print(E.symbol().spectral_radius())
+print((E.symbol().spectral_radius()))
```
Comment (hrittich) [inline demo/block_rb_jac_2x2.py]: These additional parenthesis are unecessary.

File: demo/run-all.py [hunk 3]
```diff
@@ -27,8 +27,8 @@
 ]
 
 for s in scripts:
-    print('=' * 80)
-    print('Running {}'.format(s))
+    print(('=' * 80))
```
Comment (hrittich) [inline demo/run-all.py]: These additional parenthesis are unecessary.

File: lfa_lab/core/NdRange.i [hunk 4]
```diff
@@ -41,7 +41,7 @@ class stop_iteration : public std::exception { };
                 m_pos(grid.begin())
             { }
 
-            ArrayFi next() {
+            ArrayFi __next__() {
```
Comment (hrittich) [inline lfa_lab/core/NdRange.i:44]: Does this still work in Python 2?

File: lfa_lab/report.py [hunk 5]
```diff
@@ -37,16 +37,16 @@ def print_report(E, title=''):
     if len(title) > 0:
         print('')
         print(title)
-        print('=' * len(title))
+        print(('=' * len(title)))
```
Comment (hrittich) [inline lfa_lab/report.py]: These additional parenthesis are unecessary.

File: .travis.yml [hunk 6]
```diff
@@ -8,23 +8,25 @@ addons:
...
+      - alias python=python3
+      - cmake -DCMAKE_CXX_COMPILER=g++-6 -DPYTHON_EXECUTABLE=`which python3` .
       - make
     # Out of source build
     - script:
+      - alias python=python3
       - mkdir build
       - cd build
-      - cmake -DCMAKE_CXX_COMPILER=g++-6 ../
+      - cmake -DCMAKE_CXX_COMPILER=g++-6 -DPYTHON_EXECUTABLE=`which python3` ../
```
Comment (hrittich) [inline .travis.yml]: Instead of replacing Python 2 with Python 3, I would prefer to keep support for Python 2, at least for now. Hence, it would be good to make an additional Python 3 test here, instead of changing the old one.

File: .travis.yml [hunk 7]
```diff
@@ -8,23 +8,25 @@ addons:
     packages:
     - g++-6
     - cmake
-    - python-dev
```
Comment (hrittich) [inline .travis.yml]: This should be kept. See comment below.

File: lfa_lab/core/CMakeLists.txt [hunk 8]
```diff
@@ -82,7 +82,7 @@ add_custom_command(
         ${CMAKE_CURRENT_BINARY_DIR}/extension.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/extension.py
     COMMAND
-        ${SWIG_EXECUTABLE} ARGS -c++ -python
```
Comment (hrittich) [inline lfa_lab/core/CMakeLists.txt]: This should depend on the version of python that is used.

File: lfa_lab/core/NdRange.i [hunk 9]
```diff
@@ -19,7 +19,7 @@
 
 // vim: set filetype=cpp:
 
-%exception PyNdRangeIterator::next {
```
Comment (hrittich) [inline lfa_lab/core/NdRange.i:22]: Will this change also work in Python 2 or do we need both?

File: lfa_lab/util.py [hunk 10]
```diff
@@ -165,16 +162,16 @@ def map(self, f):
         # map over a nested list (tree) of depth d
         def map_tree(d, entries):
             if d == 1:
-                return map(f, entries)
+                return list(map(f, entries))
```
Comment (hrittich) [inline lfa_lab/util.py]: Is this change really necessary?

File: lfa_lab/util.py [hunk 11]
```diff
@@ -165,16 +162,16 @@ def map(self, f):
...
             else:
-                return map(lambda e: map_tree(d-1, e), entries)
+                return [map_tree(d-1, e) for e in entries]
```
Comment (hrittich) [inline lfa_lab/util.py]: Is this change necessary?

File: lfa_lab/util.py [hunk 12]
```diff
@@ -165,16 +162,16 @@ def map(self, f):
...
         return map_tree(self.dim, self._entries)
 
 
 
 def _test():
     a = NdArray([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
-    print(a.dim)
+    print((a.dim))
```
Comment (hrittich) [inline lfa_lab/util.py]: These additional parenthesis are unecessary.

File: python_config.py [hunk 13]
```diff
@@ -10,9 +10,9 @@ def cmake_list(l):
     return ';'.join(l)
 
 def cmake_set_cache(name, value, ctype, doc):
-    print('set({0} "{1}" CACHE {2} "{3}")'.format(name, value, ctype, doc))
+    print(('set({0} "{1}" CACHE {2} "{3}")'.format(name, value, ctype, doc)))
```
Comment (hrittich) [inline python_config.py]: These additional parenthesis are unecessary.

File: python_config.py [hunk 14]
```diff
@@ -10,9 +10,9 @@ def cmake_list(l):
...
 def cmake_set(name, value):
-    print('set({0} "{1}")'.format(name, value))
+    print(('set({0} "{1}")'.format(name, value)))
```
Comment (hrittich) [inline python_config.py]: These additional parenthesis are unecessary.

File: setup.py [hunk 15]
```diff
@@ -52,7 +52,7 @@ def get_outputs(self):
             for line in in_fp:
                 outputs.append(line.strip('\n\r'))
 
-        print('Outputs returned by CMake: {}'.format(outputs))
+        print(('Outputs returned by CMake: {}'.format(outputs)))
```
Comment (hrittich) [inline setup.py]: These additional parenthesis are unecessary.

**Review by hrittich — CHANGES_REQUESTED** [review]
Thank you for your commit. I very much like to support Python 3. Since there are still many packages that only work with Python 2, I would like to keep supporting Python 2 for some time. Hence, I would like to ask you, to address the comments I have made.

Furthermore, the automatic conversion seems to have added a bit unnecessary clutter. Could you remove that? (See comments.)

File: lfa_lab/core/NdRange.i (hunk 4 above)
Comment (jonas-schmitt) [reply lfa_lab/core/NdRange.i:44]: Unfortunately not. Though, there is a way to make an iterator work both in Python 2 and 3. One can just stick for the Python two syntax (next() without underscore) and then add the following assignment to the class:
```python
__next__ = next
```
Although I'm not that familiar with SWIG and thus far have no idea how to automatically generate this additional assignment.

File: lfa_lab/core/NdRange.i (hunk 9 above)
Comment (jonas-schmitt) [reply lfa_lab/core/NdRange.i:22]: We probably need both, unless the following can be generated:
```python
__next__ = next
```

File: demo/block_jac_2x2.py (hunk 1 above)
Comment (jonas-schmitt) [reply demo/block_jac_2x2.py]: Fixed

File: demo/block_rb_jac_2x2.py (hunk 2 above)
Comment (jonas-schmitt) [reply demo/block_rb_jac_2x2.py]: Fixed

File: demo/run-all.py (hunk 3 above)
Comment (jonas-schmitt) [reply demo/run-all.py]: Fixed

File: lfa_lab/report.py (hunk 5 above)
Comment (jonas-schmitt) [reply lfa_lab/report.py]: Fixed

File: lfa_lab/util.py (hunk 10 above)
Comment (jonas-schmitt) [reply lfa_lab/util.py]: No, has been fixed.

File: lfa_lab/util.py (hunk 11 above)
Comment (jonas-schmitt) [reply lfa_lab/util.py]: No, has been fixed.

File: lfa_lab/util.py (hunk 12 above)
Comment (jonas-schmitt) [reply lfa_lab/util.py]: Fixed

File: python_config.py (hunk 13 above)
Comment (jonas-schmitt) [reply python_config.py]: Fixed

File: python_config.py (hunk 14 above)
Comment (jonas-schmitt) [reply python_config.py]: Fixed

File: setup.py (hunk 15 above)
Comment (jonas-schmitt) [reply setup.py]: Fixed

**Comment by jonas-schmitt** [comment]
I removed the unnecessary clutter that happened through the conversion. I still do not know why the Travis CI build fails for Python 3. I anyways replicated the tests for both versions. If you want to fully support both Python 2 and 3, you probably need to add an additional option to your CMakeLists.txt. I don't know if there is an elegant way to do this besides setting things manually. Furthermore, we need to address the iterator problem with SWIG.

**Comment by hrittich** [comment]
All right. I will take care of the SWIG and CMake option.

Concerning the failing travis build, have you tried an out-of-source build, i.e., a build in another directory?

**Comment by hrittich** [comment]
Rebased and merged.

Thank you very much.

**Comment by jonas-schmitt** [comment]
You're welcome. I'm glad I could help
